# Predicting sentiment from product reviews


The goal of this first notebook is to explore logistic regression and feature engineering with existing GraphLab functions.

In this notebook you will use product review data from Amazon.com to predict whether the sentiments about a product (from its reviews) are positive or negative.

* Use SFrames to do some feature engineering
* Train a logistic regression model to predict the sentiment of product reviews.
* Inspect the weights (coefficients) of a trained logistic regression model.
* Make a prediction (both class and probability) of sentiment for a new product review.
* Given the logistic regression weights, predictors and ground truth labels, write a function to compute the **accuracy** of the model.
* Inspect the coefficients of the logistic regression model and interpret their meanings.
* Compare multiple logistic regression models.

Let's get started!
    
## Fire up GraphLab Create

Make sure you have the latest version of GraphLab Create.

In [1]:
from __future__ import division
import graphlab
import math
import string

import pandas as pd
import numpy as np

# Data preparation

We will use a dataset consisting of baby product reviews on Amazon.com.

In [2]:
products = graphlab.SFrame('amazon_baby.gl/')

[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: /tmp/graphlab_server_1485316992.log


This non-commercial license of GraphLab Create for academic use is assigned to atul.acharya+dato@gmail.com and will expire on June 08, 2017.


Now, let us see a preview of what the dataset looks like.

In [3]:
products

Columns:
	name	str
	review	str
	rating	float

Rows: 183531

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|    Planetwise Flannel Wipes   | These flannel wipes are OK... |  3.0   |
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
+-------------------------------+-------------------------------+--------+
[183531 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

## Build the word count vector for each review

Let us explore a specific example of a baby product.


In [4]:
products[269]

{'name': 'The First Years Massaging Action Teether',
 'rating': 5.0,
 'review': 'A favorite in our house!'}

Now, we will perform 2 simple data transformations:

1. Remove punctuation using [Python's built-in](https://docs.python.org/2/library/string.html) string functionality.
2. Transform the reviews into word-counts.

**Aside**. In this notebook, we remove all punctuations for the sake of simplicity. A smarter approach to punctuations would preserve phrases such as "I'd", "would've", "hadn't" and so forth. See [this page](https://www.cis.upenn.edu/~treebank/tokenization.html) for an example of smart handling of punctuations.

In [5]:
import string
def remove_punctuation(text):
    return text.translate(None, string.punctuation) 

review_without_punctuation = products['review'].apply(remove_punctuation)
products['word_count'] = graphlab.text_analytics.count_words(review_without_punctuation)

Now, let us explore what the sample example above looks like after these 2 transformations. Here, each entry in the **word_count** column is a dictionary where the key is the word and the value is a count of the number of times the word occurs.

In [6]:
products[269]['word_count']

{'a': 1, 'favorite': 1, 'house': 1, 'in': 1, 'our': 1}

In [7]:
#products.show()

## Extract sentiments

We will **ignore** all reviews with *rating = 3*, since they tend to have a neutral sentiment.

In [8]:
products = products[products['rating'] != 3]
len(products)

166752

Now, we will assign reviews with a rating of 4 or higher to be *positive* reviews, while the ones with rating of 2 or lower are *negative*. For the sentiment column, we use +1 for the positive class label and -1 for the negative class label.

In [9]:
products['sentiment'] = products['rating'].apply(lambda rating : +1 if rating > 3 else -1)
products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 166752

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|     Planetwise Wipe Pouch     | it came early and was not ... |  5.0   |
| Annas Dream Full Quilt wit... | Very soft and comfortable ... |  5.0   |
| Stop Pacifier Sucking with... | This is a product well wor... |  5.0   |
| Stop Pacifier Sucking with... | All of my kids have cried ... |  5.0   |
| Stop Pacifier Sucking with... | When the Binky Fairy came ... |  5.0   |
| A Tale of Baby's Days with... | Lovely book, it's bound ti... |  4.0   |
| Baby Tracker&reg; - Daily ... | Perfect for new parents. W... |  5.0   |
| Baby Tracker&reg; - Daily ... | A friend of mine pinned th... |  5.0   |
| Baby Tracker&reg; - Daily ... | This has been an easy way ... |  4.0   |
| Baby Tracker&reg; - Daily ... | I love this journal and ou... |  4.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'and': 3, 'love': 1, 'it'... |     1     |
| {'and': 2, 'quilt': 1, 'it... |     1     |
| {'and': 3, 'ingenious': 1,... |     1     |
| {'and': 2, 'all': 2, 'help... |     1     |
| {'and': 2, 'this': 2, 'her... |     1     |
| {'shop': 1, 'noble': 1, 'i... |     1     |
| {'and': 2, 'all': 1, 'righ... |     1     |
| {'and': 1, 'fantastic': 1,... |     1     |
| {'all': 1, 'standarad': 1,... |     1     |
| {'all': 2, 'nannys': 1, 'j... |     1     |
+-------------------------------+-----------+
[166752 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

Now, we can see that the dataset contains an extra column called **sentiment** which is either positive (+1) or negative (-1).

## Split data into training and test sets

Let's perform a train/test split with 80% of the data in the training set and 20% of the data in the test set. We use `seed=1` so that everyone gets the same result.

In [10]:
train_data, test_data = products.random_split(.8, seed=1)
print len(train_data)
print len(test_data)

133416
33336


# Train a sentiment classifier with logistic regression

We will now use logistic regression to create a sentiment classifier on the training data. This model will use the column **word_count** as a feature and the column **sentiment** as the target. We will use `validation_set=None` to obtain same results as everyone else.

**Note:** This line may take 1-2 minutes.

In [11]:
sentiment_model = graphlab.logistic_classifier.create(train_data,
                                                      target = 'sentiment',
                                                      features=['word_count'],
                                                      validation_set=None)

Logistic regression:

--------------------------------------------------------

Number of examples          : 133416

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 121712

Number of coefficients    : 121713

Starting L-BFGS

--------------------------------------------------------

+-----------+----------+-----------+--------------+-------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training-accuracy |

+-----------+----------+-----------+--------------+-------------------+

| 1         | 5        | 0.000002  | 2.194093     | 0.840754          |

| 2         | 9        | 3.000000  | 3.243137     | 0.931350          |

| 3         | 10       | 3.000000  | 3.592524     | 0.882046          |

| 4         | 11       | 3.000000  | 3.962416     | 0.954076          |

| 5         | 12       | 3.000000  | 4.315573     | 0.960964          |

| 6         | 13       | 3.000000  | 4.698960     | 0.975033          |

+-----------+----------+-----------+--------------+-------------------+

TERMINATED: Terminated due to numerical difficulties.

This model may not be ideal. To improve it, consider doing one of the following:
(a) Increasing the regularization.
(b) Standardizing the input data.
(c) Removing highly correlated features.
(d) Removing `inf` and `NaN` values in the training data.

In [12]:
sentiment_model

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 121713
Number of examples             : 133416
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 121712

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.01

Training Summary
----------------
Solver                         : lbfgs
Solver iterations              : 6
Solver status                  : TERMINATED: Terminated due to numerical difficulties.
Training time (sec)            : 5.0604

Settings
--------
Log-likelihood                 : inf

Highest Positive Coefficients
-----------------------------
word_count[mobileupdate]       : 41.9847
word_count[placeid]            : 41.7354
word_count[labelbox]           : 41.151
word_count[httpwwwamazoncomreviewrhgg6qp7tdnhbrefcmcrprcmtieutf8asinb00318cla0nodeid] : 40.0454
word_count[knobskeeping]       : 36.2091

Lowest Negative Coeffi

**Aside**. You may get a warning to the effect of "Terminated due to numerical difficulties --- this model may not be ideal". It means that the quality metric (to be covered in Module 3) failed to improve in the last iteration of the run. The difficulty arises as the sentiment model puts too much weight on extremely rare words. A way to rectify this is to apply regularization, to be covered in Module 4. Regularization lessens the effect of extremely rare words. For the purpose of this assignment, however, please proceed with the model above.

Now that we have fitted the model, we can extract the weights (coefficients) as an SFrame as follows:

In [13]:
weights = sentiment_model.coefficients
weights.column_names()

['name', 'index', 'class', 'value', 'stderr']

There are a total of `121713` coefficients in the model. Recall from the lecture that positive weights $w_j$ correspond to weights that cause positive sentiment, while negative weights correspond to negative sentiment. 

Fill in the following block of code to calculate how many *weights* are positive ( >= 0). (**Hint**: The `'value'` column in SFrame *weights* must be positive ( >= 0)).

In [14]:
weights

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 121713

Data:
+-------------+-----------+-------+------------------+--------+
|     name    |   index   | class |      value       | stderr |
+-------------+-----------+-------+------------------+--------+
| (intercept) |    None   |   1   |  1.30337080544   |  None  |
|  word_count | recommend |   1   |  0.303815600015  |  None  |
|  word_count |   moist   |   1   |  0.671556821415  |  None  |
|  word_count |  osocozy  |   1   |  0.426326525702  |  None  |
|  word_count |    keps   |   1   |  7.39633708721   |  None  |
|  word_count |    leak   |   1   |  -0.24658014554  |  None  |
|  word_count |   holder  |   1   | -0.0300523581013 |  None  |
|  word_count |    was    |   1   | -0.0530004786379 |  None  |
|  word_count |    now    |   1   | 0.0383787882079  |  None  |
|  word_count |    wipe   |   1   |  0.165506649337  |  None  |
+-------------+-----------+-------+------------------+--------+
[121713 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [15]:
pos_wts = weights[ weights['value'] >= 0 ]
neg_wts = weights[ weights['value'] < 0 ]
num_positive_weights = len(pos_wts)
num_negative_weights = len(neg_wts)

print ("Number of positive weights: ",  num_positive_weights)
print ("Number of negative weights: ",  num_negative_weights)

('Number of positive weights: ', 68419)
('Number of negative weights: ', 53294)


**Quiz Question:** How many weights are >= 0?

## Making predictions with logistic regression

Now that a model is trained, we can make predictions on the **test data**. In this section, we will explore this in the context of 3 examples in the test dataset.  We refer to this set of 3 examples as the **sample_test_data**.

In [16]:
sample_test_data = test_data[10:13]
print sample_test_data['rating']
sample_test_data

[5.0, 2.0, 1.0]


Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int

Rows: 3

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|   Our Baby Girl Memory Book   | Absolutely love it and all... |  5.0   |
| Wall Decor Removable Decal... | Would not purchase again o... |  2.0   |
| New Style Trailing Cherry ... | Was so excited to get this... |  1.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+
|           word_count          | sentiment |
+-------------------------------+-----------+
| {'and': 2, 'all': 1, 'love... |     1     |
| {'and': 1, 'would': 2, 'al... |     -1    |
| {'all': 1, 'money': 1, 'in... |     -1    |
+-------------------------------+-----------+
[3 rows x 5 columns]

Let's dig deeper into the first row of the **sample_test_data**. Here's the full review:

In [17]:
sample_test_data[0]['review']

'Absolutely love it and all of the Scripture in it.  I purchased the Baby Boy version for my grandson when he was born and my daughter-in-law was thrilled to receive the same book again.'

That review seems pretty positive.

Now, let's see what the next row of the **sample_test_data** looks like. As we could guess from the sentiment (-1), the review is quite negative.

In [18]:
sample_test_data[1]['review']

'Would not purchase again or recommend. The decals were thick almost plastic like and were coming off the wall as I was applying them! The would NOT stick! Literally stayed stuck for about 5 minutes then started peeling off.'

We will now make a **class** prediction for the **sample_test_data**. The `sentiment_model` should predict **+1** if the sentiment is positive and **-1** if the sentiment is negative. Recall from the lecture that the **score** (sometimes called **margin**) for the logistic regression model  is defined as:

$$
\mbox{score}_i = \mathbf{w}^T h(\mathbf{x}_i)
$$ 

where $h(\mathbf{x}_i)$ represents the features for example $i$.  We will write some code to obtain the **scores** using GraphLab Create. For each row, the **score** (or margin) is a number in the range **[-inf, inf]**.

In [19]:
scores = sentiment_model.predict(sample_test_data, output_type='margin')
print scores

[6.73461972706021, -5.7341309967608485, -14.66846040446922]


### Predicting sentiment

These scores can be used to make class predictions as follows:

$$
\hat{y} = 
\left\{
\begin{array}{ll}
      +1 & \mathbf{w}^T h(\mathbf{x}_i) > 0 \\
      -1 & \mathbf{w}^T h(\mathbf{x}_i) \leq 0 \\
\end{array} 
\right.
$$

Using scores, write code to calculate $\hat{y}$, the class predictions:

In [20]:
# yhat = +1 if score  > 0; else -1
[1.0 if s > 0 else -1. for s in scores]

[1.0, -1.0, -1.0]

Run the following code to verify that the class predictions obtained by your calculations are the same as that obtained from GraphLab Create.

In [21]:
print "Class predictions according to GraphLab Create:" 
print sentiment_model.predict(sample_test_data)

Class predictions according to GraphLab Create:
[1, -1, -1]


**Checkpoint**: Make sure your class predictions match with the one obtained from GraphLab Create.

### Probability predictions

Recall from the lectures that we can also calculate the probability predictions from the scores using:
$$
P(y_i = +1 | \mathbf{x}_i,\mathbf{w}) = \frac{1}{1 + \exp(-\mathbf{w}^T h(\mathbf{x}_i))}.
$$

Using the variable **scores** calculated previously, write code to calculate the probability that a sentiment is positive using the above formula. For each row, the probabilities should be a number in the range **[0, 1]**.

In [22]:
import numpy as np
def sigmoid(z):
    return 1. / ( 1 + np.exp(-z))

probs = [sigmoid(z) for z in scores]
probs

[0.99881238483772095, 0.0032232681817987827, 4.2615579966524306e-07]

**Checkpoint**: Make sure your probability predictions match the ones obtained from GraphLab Create.

In [23]:
print "Class predictions according to GraphLab Create:" 
print sentiment_model.predict(sample_test_data, output_type='probability')

Class predictions according to GraphLab Create:
[0.9988123848377208, 0.0032232681817987814, 4.261557996652428e-07]


** Quiz Question:** Of the three data points in **sample_test_data**, which one (first, second, or third) has the **lowest probability** of being classified as a positive review?

# Find the most positive (and negative) review

We now turn to examining the full test dataset, **test_data**, and use GraphLab Create to form predictions on all of the test data points for faster performance.

Using the `sentiment_model`, find the 20 reviews in the entire **test_data** with the **highest probability** of being classified as a **positive review**. We refer to these as the "most positive reviews."

To calculate these top-20 reviews, use the following steps:
1.  Make probability predictions on **test_data** using the `sentiment_model`. (**Hint:** When you call `.predict` to make predictions on the test data, use option `output_type='probability'` to output the probability rather than just the most likely class.)
2.  Sort the data according to those predictions and pick the top 20. (**Hint:** You can use the `.topk` method on an SFrame to find the top k rows sorted according to the value of a specified column.)

In [24]:
# get probs for all reviews
output_test_proba = sentiment_model.predict(test_data, output_type='probability')
test_data['proba'] = output_test_proba

In [25]:
test_data[10:30]

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	proba	float

Rows: 20

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|   Our Baby Girl Memory Book   | Absolutely love it and all... |  5.0   |
| Wall Decor Removable Decal... | Would not purchase again o... |  2.0   |
| New Style Trailing Cherry ... | Was so excited to get this... |  1.0   |
| Newborn Baby Tracker&reg; ... | This is the best way to ke... |  5.0   |
| Large Jungle Animals (Pand... | I ordered these for my 7 m... |  5.0   |
| Edu-Petit Happy Jungle Day... | My 2 year old grandson got... |  5.0   |
| Modern House Giraffe Child... | It is now haning in my gra... |  5.0   |
| Modern House Safari Animal... | the instructions were real... |  5.0   |
| Neurosmith - Music Blocks ... | Our whole family loved the... |  4.0   |
| Milestones, Kids' Step Sto... | Product was very easy to u... |  5.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+-------------------+
|           word_count          | sentiment |       proba       |
+-------------------------------+-----------+-------------------+
| {'and': 2, 'all': 1, 'love... |     1     |   0.998812384838  |
| {'and': 1, 'would': 2, 'al... |     -1    |  0.0032232681818  |
| {'all': 1, 'money': 1, 'in... |     -1    | 4.26155799665e-07 |
| {'and': 1, 'this': 2, 'inf... |     1     |   0.999041882366  |
| {'and': 2, 'because': 1, '... |     1     |   0.99873091043   |
| {'and': 2, 'old': 1, 'momm... |     1     |   0.990679635875  |
| {'granddaughters': 1, 'now... |     1     |   0.999840626477  |
| {'and': 1, 'crazy': 1, 'be... |     1     |   0.131644416736  |
| {'and': 2, 'accessible': 1... |     1     |  0.0666033706634  |
| {'and': 1, 'cute': 1, 'fat... |     1     |   0.999903109199  |
+-------------------------------+-----------+-------------------+
[20 rows x 6 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [53]:
top20_products = test_data.topk('proba', k=20)
top20_products

Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	proba	float
	word_count_subset	dict

Rows: 20

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
|   Munchkin Mozart Magic Cube  | My wife and I have been li... |  4.0   |
|  BABYBJORN Potty Chair - Red  | Our family is just startin... |  5.0   |
| Safety 1st Tot-Lok Four Lo... | I have a wooden desk that ... |  5.0   |
| Summer Infant Complete Nur... | This Nursery and Bath Care... |  4.0   |
| Leachco Snoogle Total Body... | I have had my Snoogle for ... |  5.0   |
| HALO SleepSack Micro-Fleec... | I love the Sleepsack weara... |  5.0   |
| Peg Perego Primo Viaggio C... | We have been using this se... |  5.0   |
|   Capri Stroller - Red Tech   | First let me say that I wa... |  4.0   |
| Wizard Convertible Car Sea... | My son was born big and re... |  5.0   |
| Britax Marathon Convertibl... | My son began using the Mar... |  5.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+-------+
|           word_count          | sentiment | proba |
+-------------------------------+-----------+-------+
| {'rating': 1, 'all': 2, 'b... |     1     |  1.0  |
| {'managed': 1, 'just': 3, ... |     1     |  1.0  |
| {'saying': 1, 'all': 1, 'o... |     1     |  1.0  |
| {'son': 1, 'duty': 1, 'hel... |     1     |  1.0  |
| {'saying': 1, 'help': 1, '... |     1     |  1.0  |
| {'all': 1, 'just': 2, 'foo... |     1     |  1.0  |
| {'all': 1, 'just': 2, 'mov... |     1     |  1.0  |
| {'son': 3, 'infant': 1, 'a... |     1     |  1.0  |
| {'son': 4, 'infant': 1, 'd... |     1     |  1.0  |
| {'son': 3, 'infant': 1, 'o... |     1     |  1.0  |
+-------------------------------+-----------+-------+
+-------------------------------+
|       word_count_subset       |
+-------------------------------+
| {'old': 1, 'little': 1, 'w... |
| {'great': 2, 'love': 1, 'w... |
|   {'little': 2, 'would': 1}   |
| {'little': 1, 'well': 2, '... |
| {'even': 1, 'perfect': 1, ... |
| {'even': 1, 'product': 2, ... |
| {'even': 1, 'perfect': 1, ... |
| {'even': 1, 'little': 3, '... |
| {'even': 1, 'great': 2, 'l... |
| {'little': 2, 'old': 2, 'c... |
+-------------------------------+
[20 rows x 7 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [54]:
for i in range(len(top20_products['name'])):
    print(top20_products['name'][i])


Munchkin Mozart Magic Cube
BABYBJORN Potty Chair - Red
Safety 1st Tot-Lok Four Lock Assembly
Summer Infant Complete Nursery Care Kit
Leachco Snoogle Total Body Pillow
HALO SleepSack Micro-Fleece Wearable Blanket, Soft Pink, Small
Peg Perego Primo Viaggio Car Seat / Infant Carrier with LATCH Base - Black Sable
Capri Stroller - Red Tech
Wizard Convertible Car Seat with LATCH in Midnight Print
Britax Marathon Convertible Car Seat, Granite
Britax Decathlon Convertible Car Seat, Tiffany
Fisher-Price Deluxe Jumperoo
Lilly Gold Sit 'n' Stroll 5 in 1 Car Seat and Stroller Combination, Tuxedo Black (sunshade is not included in the offering)
JP Lizzy Chocolate Ice Classic Tote Set
Cloud b Sound Machine Soother, Sleep Sheep
Traveling Toddler Car Seat Travel Accessory
Ameda Purely Yours Breast Pump - Carry All
Moby Wrap Original 100% Cotton Baby Carrier, Red
Moby Wrap Original 100% Cotton Baby Carrier, Red
bumGenius One-Size Cloth Diaper Twilight


**Quiz Question**: Which of the following products are represented in the 20 most positive reviews? [multiple choice]


Now, let us repeat this exercise to find the "most negative reviews." Use the prediction probabilities to find the  20 reviews in the **test_data** with the **lowest probability** of being classified as a **positive review**. Repeat the same steps above but make sure you **sort in the opposite order**.

In [27]:
bottom20_products = test_data.topk('proba', k=20, reverse=True)
bottom20_products


Columns:
	name	str
	review	str
	rating	float
	word_count	dict
	sentiment	int
	proba	float

Rows: 20

Data:
+-------------------------------+-------------------------------+--------+
|              name             |             review            | rating |
+-------------------------------+-------------------------------+--------+
| Jolly Jumper Arctic Sneak ... | I am a "research-aholic" i... |  5.0   |
| Levana Safe N'See Digital ... | This is the first review I... |  1.0   |
| Snuza Portable Baby Moveme... | I would have given the pro... |  1.0   |
| Fisher-Price Ocean Wonders... | We have not had ANY luck w... |  2.0   |
| VTech Communications Safe ... | This is my second video mo... |  1.0   |
| Safety 1st High-Def Digita... | We bought this baby monito... |  1.0   |
| Chicco Cortina KeyFit 30 T... | My wife and I have used th... |  1.0   |
| Prince Lionheart Warmies W... | *****IMPORTANT UPDATE*****... |  1.0   |
| Valco Baby Tri-mode Twin S... | I give one star to the dim... |  1.0   |
| Adiri BPA Free Natural Nur... | I will try to write an obj... |  2.0   |
+-------------------------------+-------------------------------+--------+
+-------------------------------+-----------+--------------------+
|           word_count          | sentiment |       proba        |
+-------------------------------+-----------+--------------------+
| {'raining': 1, 'all': 8, '... |     1     | 7.80415068211e-100 |
| {'all': 2, 'they': 4, 'jus... |     -1    | 6.83650885511e-25  |
| {'contacted': 1, 'being': ... |     -1    | 2.12654510822e-24  |
| {'fishstarfish': 1, 'infan... |     -1    | 2.24582080778e-23  |
| {'all': 4, 'reviewers': 1,... |     -1    | 1.32962966148e-22  |
| {'all': 3, 'being': 1, 'ov... |     -1    | 2.06872097469e-20  |
| {'all': 4, 'wrestle': 1, '... |     -1    | 5.93881994668e-20  |
| {'less': 1, 'move': 1, 'no... |     -1    | 6.28510016533e-20  |
| {'limited': 2, 'forget': 1... |     -1    | 8.05528712681e-20  |
| {'all': 2, 'forget': 1, 'g... |     -1    | 8.46521724928e-20  |
+-------------------------------+-----------+--------------------+
[20 rows x 6 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [28]:
for i in range(len(bottom20_products['name'])):
    print(bottom20_products['name'][i])


Jolly Jumper Arctic Sneak A Peek Infant Car Seat Cover Black
Levana Safe N'See Digital Video Baby Monitor with Talk-to-Baby Intercom and Lullaby Control (LV-TW501)
Snuza Portable Baby Movement Monitor
Fisher-Price Ocean Wonders Aquarium Bouncer
VTech Communications Safe &amp; Sounds Full Color Video and Audio Monitor
Safety 1st High-Def Digital Monitor
Chicco Cortina KeyFit 30 Travel System in Adventure
Prince Lionheart Warmies Wipes Warmer
Valco Baby Tri-mode Twin Stroller EX- Hot Chocolate
Adiri BPA Free Natural Nurser Ultimate Bottle Stage 1 White, Slow Flow (0-3 months)
Munchkin Nursery Projector and Sound System, White
The First Years True Choice P400 Premium Digital Monitor, 2 Parent Unit
Nuby Natural Touch Silicone Travel Infa Feeder, Colors May Vary, 3 Ounce
Peg-Perego Tatamia High Chair, White Latte
Fisher-Price Royal Potty
Safety 1st Exchangeable Tip 3 in 1 Thermometer
Safety 1st Lift Lock and Swing Gate
Evenflo Take Me Too Premiere Tandem Stroller - Castlebay
Cloth Diaper Sp

**Quiz Question**: Which of the following products are represented in the 20 most negative reviews?  [multiple choice]

## Compute accuracy of the classifier

We will now evaluate the accuracy of the trained classifier. Recall that the accuracy is given by


$$
\mbox{accuracy} = \frac{\mbox{# correctly classified examples}}{\mbox{# total examples}}
$$

This can be computed as follows:

* **Step 1:** Use the trained model to compute class predictions (**Hint:** Use the `predict` method)
* **Step 2:** Count the number of data points when the predicted class labels match the ground truth labels (called `true_labels` below).
* **Step 3:** Divide the total number of correct predictions by the total number of data points in the dataset.

Complete the function below to compute the classification accuracy:

In [29]:
def get_classification_accuracy(model, data, true_labels):
    # First get the predictions
    ## YOUR CODE HERE
    predictions = model.predict(data)
    
    # Compute the number of correctly classified examples
    ## YOUR CODE HERE
    diff = predictions - true_labels
    num_correct = len(diff[diff == 0])

    # Then compute accuracy by dividing num_correct by total number of examples
    ## YOUR CODE HERE
    accuracy = num_correct / len(true_labels)
    
    return accuracy

Now, let's compute the classification accuracy of the **sentiment_model** on the **test_data**.

In [30]:
get_classification_accuracy(sentiment_model, test_data, test_data['sentiment'])

0.9145368370530358

**Quiz Question**: What is the accuracy of the **sentiment_model** on the **test_data**? Round your answer to 2 decimal places (e.g. 0.76).

**Quiz Question**: Does a higher accuracy value on the **training_data** always imply that the classifier is better?

## Learn another classifier with fewer words

There were a lot of words in the model we trained above. We will now train a simpler logistic regression model using only a subset of words that occur in the reviews. For this assignment, we selected a 20 words to work with. These are:

In [31]:
significant_words = ['love', 'great', 'easy', 'old', 'little', 'perfect', 'loves', 
      'well', 'able', 'car', 'broke', 'less', 'even', 'waste', 'disappointed', 
      'work', 'product', 'money', 'would', 'return']

In [32]:
len(significant_words)

20

For each review, we will use the **word_count** column and trim out all words that are **not** in the **significant_words** list above. We will use the [SArray dictionary trim by keys functionality]( https://dato.com/products/create/docs/generated/graphlab.SArray.dict_trim_by_keys.html). Note that we are performing this on both the training and test set.

In [33]:
train_data['word_count_subset'] = train_data['word_count'].dict_trim_by_keys(significant_words, exclude=False)
test_data['word_count_subset']  = test_data['word_count'].dict_trim_by_keys(significant_words, exclude=False)

Let's see what the first example of the dataset looks like:

In [34]:
train_data[0]['review']

'it came early and was not disappointed. i love planet wise bags and now my wipe holder. it keps my osocozy wipes moist and does not leak. highly recommend it.'

The **word_count** column had been working with before looks like the following:

In [35]:
print train_data[0]['word_count']

{'and': 3, 'love': 1, 'it': 3, 'highly': 1, 'osocozy': 1, 'bags': 1, 'holder': 1, 'leak': 1, 'moist': 1, 'does': 1, 'recommend': 1, 'was': 1, 'wipes': 1, 'early': 1, 'not': 2, 'now': 1, 'disappointed': 1, 'wipe': 1, 'keps': 1, 'wise': 1, 'i': 1, 'planet': 1, 'my': 2, 'came': 1}


Since we are only working with a subset of these words, the column **word_count_subset** is a subset of the above dictionary. In this example, only 2 `significant words` are present in this review.

In [36]:
print train_data[0]['word_count_subset']

{'love': 1, 'disappointed': 1}


## Train a logistic regression model on a subset of data

We will now build a classifier with **word_count_subset** as the feature and **sentiment** as the target. 

In [37]:
simple_model = graphlab.logistic_classifier.create(train_data,
                                                   target = 'sentiment',
                                                   features=['word_count_subset'],
                                                   validation_set=None)
simple_model

Logistic regression:

--------------------------------------------------------

Number of examples          : 133416

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 20

Number of coefficients    : 21

Starting Newton Method

--------------------------------------------------------

+-----------+----------+--------------+-------------------+

| Iteration | Passes   | Elapsed Time | Training-accuracy |

+-----------+----------+--------------+-------------------+

| 1         | 2        | 0.117070     | 0.862917          |

| 2         | 3        | 0.183979     | 0.865713          |

| 3         | 4        | 0.250247     | 0.866478          |

| 4         | 5        | 0.324682     | 0.866748          |

| 5         | 6        | 0.405775     | 0.866815          |

| 6         | 7        | 0.475180     | 0.866815          |

+-----------+----------+--------------+-------------------+

SUCCESS: Optimal solution found.

Class                          : LogisticClassifier

Schema
------
Number of coefficients         : 21
Number of examples             : 133416
Number of classes              : 2
Number of feature columns      : 1
Number of unpacked features    : 20

Hyperparameters
---------------
L1 penalty                     : 0.0
L2 penalty                     : 0.01

Training Summary
----------------
Solver                         : newton
Solver iterations              : 6
Solver status                  : SUCCESS: Optimal solution found.
Training time (sec)            : 0.489

Settings
--------
Log-likelihood                 : 44323.7254

Highest Positive Coefficients
-----------------------------
word_count_subset[loves]       : 1.6773
word_count_subset[perfect]     : 1.5145
word_count_subset[love]        : 1.3654
(intercept)                    : 1.2995
word_count_subset[easy]        : 1.1937

Lowest Negative Coefficients
----------------------------
word_count_subset[disappointed] : -2.3551
wor

We can compute the classification accuracy using the `get_classification_accuracy` function you implemented earlier.

In [38]:
get_classification_accuracy(simple_model, test_data, test_data['sentiment'])

0.8693004559635229

Now, we will inspect the weights (coefficients) of the **simple_model**:

In [39]:
simple_model.coefficients

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 21

Data:
+-------------------+--------------+-------+-----------------+-----------------+
|        name       |    index     | class |      value      |      stderr     |
+-------------------+--------------+-------+-----------------+-----------------+
|    (intercept)    |     None     |   1   |   1.2995449552  | 0.0120888541331 |
| word_count_subset | disappointed |   1   |  -2.35509250061 | 0.0504149888557 |
| word_count_subset |     love     |   1   |  1.36543549368  | 0.0303546295109 |
| word_count_subset |    little    |   1   |  0.520628636025 | 0.0214691475665 |
| word_count_subset |    loves     |   1   |  1.67727145556  | 0.0482328275384 |
| word_count_subset |   product    |   1   | -0.320555492996 | 0.0154311321362 |
| word_count_subset |     well     |   1   |  0.504256746398 |  0.021381300631 |
| word_count_subset |    great     |   1   |  0.94469126948  | 0.0209509926591 |
| word_count_subset |     easy     |   1   |  1.19366189833  |  0.029288869202 |
| word_count_subset |     work     |   1   | -0.621700012425 | 0.0230330597946 |
+-------------------+--------------+-------+-----------------+-----------------+
[21 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

Let's sort the coefficients (in descending order) by the **value** to obtain the coefficients with the most positive effect on the sentiment.

In [40]:
simple_model.coefficients.sort('value', ascending=False).print_rows(num_rows=21)

+-------------------+--------------+-------+-----------------+-----------------+
|        name       |    index     | class |      value      |      stderr     |
+-------------------+--------------+-------+-----------------+-----------------+
| word_count_subset |    loves     |   1   |  1.67727145556  | 0.0482328275384 |
| word_count_subset |   perfect    |   1   |  1.51448626703  |  0.049861952294 |
| word_count_subset |     love     |   1   |  1.36543549368  | 0.0303546295109 |
|    (intercept)    |     None     |   1   |   1.2995449552  | 0.0120888541331 |
| word_count_subset |     easy     |   1   |  1.19366189833  |  0.029288869202 |
| word_count_subset |    great     |   1   |  0.94469126948  | 0.0209509926591 |
| word_count_subset |    little    |   1   |  0.520628636025 | 0.0214691475665 |
| word_count_subset |     well     |   1   |  0.504256746398 |  0.021381300631 |
| word_count_subset |     able     |   1   |  0.191438302295 | 0.0337581955697 |
| word_count_subset |     ol

In [41]:
simple_model.coefficients[3]

{'class': 1,
 'index': 'little',
 'name': 'word_count_subset',
 'stderr': 0.021469147566490373,
 'value': 0.5206286360250183}

In [42]:
cf_words = []
cf_vals = []
for i in range(len(simple_model.coefficients)):
    word = simple_model.coefficients[i]['index']
    cf_words.append(word)
    val = simple_model.coefficients[i]['value']
    cf_vals.append(val)

#cf_vals

In [43]:
cf = pd.DataFrame()

In [44]:
cf['word'] = cf_words
cf['vals'] = cf_vals

cf

word      vals
0           None  1.299545
1   disappointed -2.355093
2           love  1.365435
3         little  0.520629
4          loves  1.677271
5        product -0.320555
6           well  0.504257
7          great  0.944691
8           easy  1.193662
9           work -0.621700
10         would -0.362309
11          able  0.191438
12       perfect  1.514486
13           old  0.085396
14           car  0.058835
15         money -0.897884
16         waste -2.042774
17          even -0.511739
18          less -0.209710
19         broke -1.657964
20        return -2.117297

In [45]:
# positive words only
cf[cf['vals'] > 0]

word      vals
0      None  1.299545
2      love  1.365435
3    little  0.520629
4     loves  1.677271
6      well  0.504257
7     great  0.944691
8      easy  1.193662
11     able  0.191438
12  perfect  1.514486
13      old  0.085396
14      car  0.058835

**Quiz Question**: Consider the coefficients of **simple_model**. There should be 21 of them, an intercept term + one for each word in **significant_words**. How many of the 20 coefficients (corresponding to the 20 **significant_words** and *excluding the intercept term*) are positive for the `simple_model`?

In [114]:
## ANSWER: 10 of these words from simple_model are Positive

**Quiz Question**: Are the positive words in the **simple_model** (let us call them `positive_significant_words`) also positive words in the **sentiment_model**?

In [115]:
sentiment_model.coefficients

Columns:
	name	str
	index	str
	class	int
	value	float
	stderr	float

Rows: 121713

Data:
+-------------+-----------+-------+------------------+--------+
|     name    |   index   | class |      value       | stderr |
+-------------+-----------+-------+------------------+--------+
| (intercept) |    None   |   1   |  1.30337080544   |  None  |
|  word_count | recommend |   1   |  0.303815600015  |  None  |
|  word_count |   moist   |   1   |  0.671556821415  |  None  |
|  word_count |  osocozy  |   1   |  0.426326525702  |  None  |
|  word_count |    keps   |   1   |   7.3963370872   |  None  |
|  word_count |    leak   |   1   |  -0.24658014554  |  None  |
|  word_count |   holder  |   1   | -0.0300523581013 |  None  |
|  word_count |    was    |   1   | -0.0530004786379 |  None  |
|  word_count |    now    |   1   | 0.0383787882079  |  None  |
|  word_count |    wipe   |   1   |  0.165506649337  |  None  |
+-------------+-----------+-------+------------------+--------+
[121713 rows x 5 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [ ]:
ww = []   # words
vv = []   # values
for i in range(len(sentiment_model.coefficients)):
    ww.append(sentiment_model.coefficients[i]['index'])
    vv.append(sentiment_model.coefficients[i]['value'])

sentiment_cf = pd.DataFrame()
sentiment_cf['words'] = ww
sentiment_cf['value'] = vv

sentiment_cf_pos = sentiment_cf['value' > 0]
len(sentiment_cf_pos)

In [ ]:
sentiment_model.coefficients.sort('value', ascending=False).print_rows(num_rows=21)

# Comparing models

We will now compare the accuracy of the **sentiment_model** and the **simple_model** using the `get_classification_accuracy` method you implemented above.

First, compute the classification accuracy of the **sentiment_model** on the **train_data**:

In [46]:
acc_senti = get_classification_accuracy(sentiment_model, train_data, train_data['sentiment'])
print(acc_senti)

0.979440247047


Now, compute the classification accuracy of the **simple_model** on the **train_data**:

In [47]:
acc_simple = get_classification_accuracy(simple_model, train_data, train_data['sentiment'])
print(acc_simple)

0.866815074654


**Quiz Question**: Which model (**sentiment_model** or **simple_model**) has higher accuracy on the TRAINING set?

Now, we will repeat this exercise on the **test_data**. Start by computing the classification accuracy of the **sentiment_model** on the **test_data**:

In [48]:
# acc on test data
acc_senti_test = get_classification_accuracy(sentiment_model, test_data, test_data['sentiment'])
print(acc_senti_test)

0.914536837053


Next, we will compute the classification accuracy of the **simple_model** on the **test_data**:

In [49]:
acc_simple_test = get_classification_accuracy(simple_model, test_data, test_data['sentiment'])
print(acc_simple_test)

0.869300455964


**Quiz Question**: Which model (**sentiment_model** or **simple_model**) has higher accuracy on the TEST set?

## Baseline: Majority class prediction

It is quite common to use the **majority class classifier** as the a baseline (or reference) model for comparison with your classifier model. The majority classifier model predicts the majority class for all data points. At the very least, you should healthily beat the majority class classifier, otherwise, the model is (usually) pointless.

What is the majority class in the **train_data**?

In [50]:
num_positive  = (train_data['sentiment'] == +1).sum()
num_negative = (train_data['sentiment'] == -1).sum()
print num_positive
print num_negative

112164
21252


Now compute the accuracy of the majority class classifier on **test_data**.

**Quiz Question**: Enter the accuracy of the majority class classifier model on the **test_data**. Round your answer to two decimal places (e.g. 0.76).

In [51]:
num_positive_test  = (test_data['sentiment'] == +1).sum()
num_negative_test = (test_data['sentiment'] == -1).sum()
print num_positive_test
print num_negative_test
print num_positive_test / (num_positive_test + num_negative_test)

28095
5241
0.842782577394


**Quiz Question**: Is the **sentiment_model** definitely better than the majority class classifier (the baseline)?